### PR TITLE
Add class-balanced sampling and loss weighting

### PR DESCRIPTION
**Pull request by ShuaiLYU**
This pull request introduces class-balanced sampling and loss weighting to improve training performance, especially on imbalanced datasets. The main changes add a `BalancedDistributedSampler` for class-balanced data loading, support for per-class loss balancing, and corresponding configuration options. These features help ensure each class is well represented during training and that the loss function accounts for class imbalance.

**Class-balanced data loading and loss weighting:**

* Added `BalancedDistributedSampler` to `ultralytics/data/build.py` for strict per-class balanced sampling across distributed training, ensuring each class contributes equally to each epoch.
* Modified `build_dataloader` to support a new `balance_sampler` argument and logic to select the balanced sampler when enabled. [[1]](diffhunk://#diff-e8bb4966653e657a3bfc4a4c892ccdcf3c486a982016d2e51631b5b9393c2c5fR510) [[2]](diffhunk://#diff-e8bb4966653e657a3bfc4a4c892ccdcf3c486a982016d2e51631b5b9393c2c5fR522) [[3]](diffhunk://#diff-e8bb4966653e657a3bfc4a4c892ccdcf3c486a982016d2e51631b5b9393c2c5fL317-R546)
* Added `balance_sampler` and `loss_balance` options to the configuration (`ultralytics/cfg/default.yaml`) to control class-balanced sampling and loss balancing.

**Class-balanced loss calculation:**

* Introduced `build_name_to_weight` in `ultralytics/data/build.py` to compute per-class weights for loss balancing based on class frequencies using various weighting schemes.
* In `ultralytics/models/yolo/detect/train.py`, implemented logic to compute and cache per-class weights (`get_dataset_cls_weight`), and inject them into the batch for loss calculation when `loss_balance` is enabled. [[1]](diffhunk://#diff-d65f53ba236420b68c8dded0a5b817050a77ec804b5975d67e448a17b0d17e0aR105-R126) [[2]](diffhunk://#diff-d65f53ba236420b68c8dded0a5b817050a77ec804b5975d67e448a17b0d17e0aR136-R140)
* Updated the loss computation in `ultralytics/utils/loss.py` to apply class weights if present in the batch, supporting both precomputed and per-batch dynamic weighting. [[1]](diffhunk://#diff-d2ac3773214d5694c3286cf9534fc4914555f1a3f5c5ae4a99c6aff00c75892dR426-R430) [[2]](diffhunk://#diff-d2ac3773214d5694c3286cf9534fc4914555f1a3f5c5ae4a99c6aff00c75892dR474-R491)

**Logging and utility improvements:**

* Integrated logging for class distributions and warnings when class names are missing from the weight mapping. [[1]](diffhunk://#diff-e8bb4966653e657a3bfc4a4c892ccdcf3c486a982016d2e51631b5b9393c2c5fR216-R431) [[2]](diffhunk://#diff-d2ac3773214d5694c3286cf9534fc4914555f1a3f5c5ae4a99c6aff00c75892dL16-R16)

These changes collectively enable robust handling of class imbalance during both data loading and loss computation, improving training effectiveness on unbalanced datasets.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ShuaiLYU, thank you for submitting a `ultralytics/ultralytics` 🚀 PR for the new class-balanced sampling feature! This is an automated message to help with review, and an engineer will assist you shortly. Please review the checklist below to help ensure smooth integration. 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good direction overall, but the new sampler has two DDP correctness issues that should be fixed before merge: ranks currently receive contiguous class blocks, and shard sizes are uneven. Both can break or materially distort multi-GPU training. Aside from that, the integration looks clean; the remaining issue is a small logging bug in the optional histogram helper.

💬 Posted 3 inline comments

File: ultralytics/data/build.py [hunk 1]
```diff
@@ -214,6 +214,204 @@ def set_epoch(self, epoch: int) -> None:
...
+            img_list = self.cls_to_imgs[cls_id]
+            n = len(img_list)
+            sampled = torch.randint(0, n, (self.samples_per_class,), generator=g).tolist()
+            indices.extend(img_list[p] for p in sampled)
+
+        # 2. Optionally save class-distribution histogram (all ranks)
+        # if self.rank == 0:
+        #     self._save_cls_histogram(indices)
+
+        # 3. Shard across ranks (contiguous slice)
```
Comment (UltralyticsAssistant) [inline ultralytics/data/build.py]: ⚠️ **HIGH**: `indices` is built class-by-class and only shuffled after the per-rank slice is taken. In distributed training that means each rank gets a contiguous block of classes rather than a representative mix, so different GPUs can spend an epoch training on mostly disjoint class subsets. That undermines the balancing goal and can skew gradients. Shuffle or interleave the global sampled index list before sharding it across ranks.

File: ultralytics/data/build.py [hunk 2]
```diff
@@ -214,6 +214,204 @@ def set_epoch(self, epoch: int) -> None:
...
+        per_rank = self.total_size // self.num_replicas
```
Comment (UltralyticsAssistant) [inline ultralytics/data/build.py]: ⚠️ **HIGH**: This sharding logic gives ranks different sample counts whenever `total_size % num_replicas != 0`, and some ranks can get zero samples when `total_size < num_replicas`. Unlike `DistributedSampler`, that can make DDP workers run a different number of steps and hang on the next collective. Please pad/truncate the sampled indices so every rank gets the same fixed `num_samples`.

File: ultralytics/data/build.py [hunk 3]
```diff
@@ -214,6 +214,204 @@ def set_epoch(self, epoch: int) -> None:
...
+        save_dir = Path("runs") / "balanced_sampler"
+        save_dir.mkdir(parents=True, exist_ok=True)
+        suffix = f"epoch{self.epoch:04d}_rank{self.rank}"
+        save_path = save_dir / f"cls_dist_{suffix}.png"
+        plt.savefig(save_path, dpi=120)
+        plt.close(fig)
+
+        
+        # Save CSV: cls_id, bbox_count — one row per class
+        LOGGER.info("Saved class distribution histogram to", save_path)
```
Comment (UltralyticsAssistant) [inline ultralytics/data/build.py]: 📝 **LOW**: `LOGGER.info` is being called with multiple positional arguments but no `%s` placeholder or f-string. With the standard logger API this raises a formatting error when `_save_cls_histogram()` is enabled. The same issue appears on the CSV log line below.

**Comment by Laughing-q** [comment]
@ShuaiLYU checkout the comments from the  bot and fix the failed tests

**Comment by ShuaiLYU** [comment]
> @ShuaiLYU checkout the comments from the bot and fix the failed tests

Done @Laughing-q

**Comment by ShuaiLYU** [comment]
@Laughing-q  did some quick sanity checks on coco128 with 3 configs:

```bash
# baseline
yolo train model=yolo11n.pt data=coco128.yaml epochs=15 batch=8 balance_sampler=False loss_balance=False
# data balancing
yolo train model=yolo11n.pt data=coco128.yaml epochs=15 batch=8 balance_sampler=True  loss_balance=False
# loss balancing
yolo train model=yolo11n.pt data=coco128.yaml epochs=15 batch=8 balance_sampler=False loss_balance=True
```


Both strategies hurt overall mAP on coco128. `balance_sampler` shows ↑ on 18/71 classes, ↓ on 48. `loss_balance` shows ↑ on 18/71, ↓ on 43.

These strategies may behave much better on larger-scale datasets (COCO, Objects365). My gpus are occupied for ablation study of yoleo26.  I will run these experiments on them when I have free gpus, and update the results here. If you have free gpus, you can also run these experiments and share the results here.
**Overall mAP50-95:**

| Config | mAP50-95 | Δ baseline |
|---|---|---|
| baseline | 0.566 | — |
| balance_sampler | 0.536 | -0.030 |
| loss_balance | 0.546 | -0.020 |



<details>
<summary>Per-class breakdown (click to expand)</summary>

|  |  |  | baseline | balance_sampler | loss_balance |
|---|---|---|---|---|---|
| **Class** | **Imgs** | **Insts** | **mAP50-95** | **mAP50-95** | **mAP50-95** |
| **all** | **128** | **929** | **0.566** | **0.536 ↓** | **0.546 ↓** |
| person | 61 | 254 | 0.569 | 0.551 ↓ | 0.561 ↓ |
| bicycle | 3 | 6 | 0.373 | 0.259 ↓ | 0.303 ↓ |
| car | 12 | 46 | 0.211 | 0.174 ↓ | 0.185 ↓ |
| motorcycle | 4 | 5 | 0.787 | 0.826 ↑ | 0.827 ↑ |
| airplane | 5 | 6 | 0.904 | 0.880 ↓ | 0.910 ↑ |
| bus | 5 | 7 | 0.701 | 0.669 ↓ | 0.676 ↓ |
| train | 3 | 3 | 0.737 | 0.841 ↑ | 0.739 ↑ |
| truck | 5 | 12 | 0.371 | 0.271 ↓ | 0.335 ↓ |
| boat | 2 | 6 | 0.441 | 0.417 ↓ | 0.429 ↓ |
| traffic light | 4 | 14 | 0.184 | 0.178 ↓ | 0.174 ↓ |
| stop sign | 2 | 2 | 0.846 | 0.797 ↓ | 0.897 ↑ |
| bench | 5 | 9 | 0.428 | 0.398 ↓ | 0.313 ↓ |
| bird | 2 | 16 | 0.671 | 0.657 ↓ | 0.699 ↑ |
| cat | 4 | 4 | 0.908 | 0.842 ↓ | 0.895 ↓ |
| dog | 9 | 9 | 0.721 | 0.545 ↓ | 0.734 ↑ |
| horse | 1 | 2 | 0.749 | 0.796 ↑ | 0.748 ↓ |
| elephant | 4 | 17 | 0.789 | 0.734 ↓ | 0.787 ↓ |
| bear | 1 | 1 | 0.995 | 0.995 = | 0.995 = |
| zebra | 2 | 4 | 0.972 | 0.971 ↓ | 0.973 ↑ |
| giraffe | 4 | 9 | 0.751 | 0.667 ↓ | 0.734 ↓ |
| backpack | 4 | 6 | 0.306 | 0.260 ↓ | 0.227 ↓ |
| umbrella | 4 | 18 | 0.568 | 0.493 ↓ | 0.513 ↓ |
| handbag | 9 | 19 | 0.235 | 0.122 ↓ | 0.142 ↓ |
| tie | 6 | 7 | 0.654 | 0.540 ↓ | 0.571 ↓ |
| suitcase | 2 | 4 | 0.697 | 0.765 ↑ | 0.631 ↓ |
| frisbee | 5 | 5 | 0.656 | 0.647 ↓ | 0.663 ↑ |
| skis | 1 | 1 | 0.697 | 0.597 ↓ | 0.796 ↑ |
| snowboard | 2 | 7 | 0.505 | 0.555 ↑ | 0.576 ↑ |
| sports ball | 6 | 6 | 0.260 | 0.291 ↑ | 0.264 ↑ |
| kite | 2 | 10 | 0.233 | 0.183 ↓ | 0.184 ↓ |
| baseball bat | 4 | 4 | 0.107 | 0.174 ↑ | 0.088 ↓ |
| baseball glove | 4 | 7 | 0.266 | 0.216 ↓ | 0.227 ↓ |
| skateboard | 3 | 5 | 0.549 | 0.528 ↓ | 0.545 ↓ |
| tennis racket | 5 | 7 | 0.381 | 0.294 ↓ | 0.386 ↑ |
| bottle | 6 | 18 | 0.330 | 0.338 ↑ | 0.335 ↑ |
| wine glass | 5 | 16 | 0.484 | 0.414 ↓ | 0.467 ↓ |
| cup | 10 | 36 | 0.353 | 0.295 ↓ | 0.321 ↓ |
| fork | 6 | 6 | 0.262 | 0.281 ↑ | 0.223 ↓ |
| knife | 7 | 16 | 0.352 | 0.392 ↑ | 0.406 ↑ |
| spoon | 5 | 22 | 0.289 | 0.283 ↓ | 0.291 ↑ |
| bowl | 9 | 28 | 0.663 | 0.582 ↓ | 0.642 ↓ |
| banana | 1 | 1 | 0.995 | 0.995 = | 0.995 = |
| sandwich | 2 | 2 | 0.770 | 0.995 ↑ | 0.547 ↓ |
| orange | 1 | 4 | 0.598 | 0.574 ↓ | 0.494 ↓ |
| broccoli | 4 | 11 | 0.314 | 0.229 ↓ | 0.238 ↓ |
| carrot | 3 | 24 | 0.552 | 0.506 ↓ | 0.523 ↓ |
| hot dog | 1 | 2 | 0.995 | 0.995 = | 0.995 = |
| pizza | 5 | 5 | 0.872 | 0.836 ↓ | 0.854 ↓ |
| donut | 2 | 14 | 0.827 | 0.793 ↓ | 0.799 ↓ |
| cake | 4 | 4 | 0.865 | 0.797 ↓ | 0.895 ↑ |
| chair | 9 | 35 | 0.367 | 0.274 ↓ | 0.318 ↓ |
| couch | 5 | 6 | 0.667 | 0.607 ↓ | 0.635 ↓ |
| potted plant | 9 | 14 | 0.599 | 0.593 ↓ | 0.599 = |
| bed | 3 | 3 | 0.819 | 0.648 ↓ | 0.863 ↑ |
| dining table | 10 | 13 | 0.532 | 0.467 ↓ | 0.487 ↓ |
| toilet | 2 | 2 | 0.795 | 0.995 ↑ | 0.796 ↑ |
| tv | 2 | 2 | 0.895 | 0.846 ↓ | 0.895 = |
| laptop | 2 | 3 | 0.479 | 0.652 ↑ | 0.437 ↓ |
| mouse | 2 | 2 | 0.032 | 0.021 ↓ | 0.014 ↓ |
| remote | 5 | 8 | 0.557 | 0.442 ↓ | 0.506 ↓ |
| cell phone | 5 | 8 | 0.211 | 0.157 ↓ | 0.166 ↓ |
| microwave | 3 | 3 | 0.930 | 0.782 ↓ | 0.930 = |
| oven | 5 | 5 | 0.420 | 0.277 ↓ | 0.349 ↓ |
| sink | 4 | 6 | 0.337 | 0.279 ↓ | 0.303 ↓ |
| refrigerator | 5 | 5 | 0.745 | 0.670 ↓ | 0.714 ↓ |
| book | 6 | 29 | 0.225 | 0.194 ↓ | 0.200 ↓ |
| clock | 8 | 9 | 0.807 | 0.766 ↓ | 0.824 ↑ |
| vase | 2 | 2 | 0.895 | 0.895 = | 0.895 = |
| scissors | 1 | 1 | 0.000 | 0.005 ↑ | 0.000 = |
| teddy bear | 6 | 21 | 0.541 | 0.422 ↓ | 0.505 ↓ |
| toothbrush | 2 | 5 | 0.579 | 0.615 ↑ | 0.541 ↓ |

</details>

**Comment by glenn-jocher** [comment]
Hi @ShuaiLYU — flagging a blocker before this can move forward. By your own benchmark ([comment](https://github.com/ultralytics/ultralytics/pull/24066#issuecomment-4174753393)), both paths regress accuracy on coco128: baseline mAP50-95 **0.566** vs `balance_sampler` **0.536** (−0.030) and `loss_balance` **0.546** (−0.020), degrading 43–48 of 71 classes. On top of that the PR adds a fair amount of debug/logging plumbing to the training path.

Per our "less is more / zero-regression" bar, we can't merge a feature that lowers mAP by its own numbers. Two ways forward:
1. If you believe the idea has legs, show a **net mAP gain on a real imbalanced benchmark** (coco128 is too small/noisy to conclude — a class-imbalanced subset of COCO or a long-tail dataset), and trim the implementation to the minimal sampler/loss change (drop the debug/logging scaffolding).
2. Otherwise, let's close it for now.

Could you either post updated results or close? Thanks!

**Comment by glenn-jocher** [comment]
Thank you @ShuaiLYU for the implementation work and especially for publishing the benchmark results so transparently—that evidence is exactly what helps us make good maintenance decisions. At the current head, both strategies regress coco128 mAP and the generic sampler/loss/debug surface is too large to retain without a demonstrated long-tail gain, so we’re closing this iteration. The idea may still be worthwhile with a representative imbalanced benchmark and a much smaller ownership-focused implementation. We value the exploration and the honest results, and we’d be happy to reconsider a focused follow-up backed by a net gain. 🙏
